File: frontend/src/services/mock/mockAuthService.ts
import type { AuthService, AuthStateListener, AuthUser } from "../auth";

export class MockAuthService implements AuthService {
  private user: AuthUser;
  private readonly listeners = new Set<AuthStateListener>();

  constructor(
    initialUser: AuthUser = {
      uid: "user-owner",
      displayName: "나",
      isAnonymous: true,
    },
  ) {
    this.user = structuredClone(initialUser);
  }

  subscribe(listener: AuthStateListener) {
    this.listeners.add(listener);
    queueMicrotask(() => listener(structuredClone(this.user)));

    return () => {
      this.listeners.delete(listener);
    };
  }

  async ensureAnonymousSession(): Promise<AuthUser> {
    this.emit();
    return structuredClone(this.user);
  }

  async linkGoogleAccount(): Promise<AuthUser> {
    this.user = {
      ...this.user,
      displayName: "나 (Google 연결)",
      email: "demo@example.com",
      isAnonymous: false,
    };
    this.emit();
    return structuredClone(this.user);
  }

  private emit() {
    for (const listener of this.listeners) {
      listener(structuredClone(this.user));
    }
  }
}

File: frontend/lib/services/mock_auth_service.dart
import 'dart:async';

import 'auth_service.dart';

final class MockAuthService implements AuthService {
  MockAuthService({
    AuthUser initialUser = const AuthUser(
      uid: 'tokyo-owner',
      displayName: '나',
      isAnonymous: true,
    ),
  }) : _user = initialUser;

  final StreamController<AuthUser?> _changes = StreamController.broadcast();
  AuthUser _user;

  @override
  Stream<AuthUser?> authStateChanges() async* {
    yield _user;
    yield* _changes.stream;
  }

  @override
  Future<AuthUser> ensureAnonymousSession() async {
    _changes.add(_user);
    return _user;
  }

  @override
  Future<AuthUser> linkGoogleAccount() async {
    _user = AuthUser(
      uid: _user.uid,
      displayName: '나 (Google 연결)',
      email: 'demo@example.com',
      photoUrl: _user.photoUrl,
      isAnonymous: false,
    );
    _changes.add(_user);
    return _user;
  }

  Future<void> dispose() => _changes.close();
}
